### PR TITLE
perf: DSA Indexer-Topk fused kernel: litetopk

### PR DESCRIPTION
**Pull request by Heisenberg-Yin**
## 📌 Description

Hi @yzh119 @bkryu @aleozlx @yongwww @yyihuang, could you please take a look at this PR when you have a chance?

This PR proposes a fused DSA Indexed-TopK kernel (LiteTopk), which never materializes the [num_q, seq_len] matrix. To the best of our knowledge, this is the first of its kind. The corresponding paper is available at https://github.com/Heisenberg-Yin/LiteTopK/blob/main/flashtopk.pdf. **LiteTopK accelerates GLM-5.2 1M-token prefill by 1.2× without accuracy loss.**

DSA models such as **GLM-4.6 / GLM-5.2, DeepSeek-V3.2, and DeepSeek-V4** run a lightweight
*indexer* before main attention: an fp8 multi-query attention scores every KV
position against each query to produce a `[num_q, seq_len]` logit matrix, and the
**top-k** KV positions per query are selected so that main attention only attends
to those. 

**At long context (256K–1M) this indexer top-k becomes a real bottleneck for both memory overhead and latency**, because the common path **materializes the full `[num_q, seq_len]` fp32 logit
matrix** in HBM and then runs a per-row top-k over it — an `O(num_q × seq_len)`
transient buffer and two passes over it.

LiteTopK avoids to store the score matrix in the following way:

1. **fp8 MQA scoring** of each KV tile against the queries (tcgen05 UMMA, using a
   vendored subset of DeepSeek's DeepGEMM SM100 primitives);
2. maintain an **online bucketed gate** that keeps only candidates above a running,
   self-calibrating threshold (256 histogram buckets), so the vast majority of
   positions are discarded on the fly;
3. a **compact top-k** over the surviving candidates.

Working memory is a **constant-size candidate buffer** (`cand_cap`, independent of
`seq_len`) instead of an `O(num_q × seq_len)` logit matrix. A **hot-start** mode
seeds the gate calibration from a `sample_logits` tensor (e.g. the previous chunk's
top-k carried forward), which tightens the threshold so even fewer candidates
survive the scan.

### API

```python
import flashinfer

# q:             [num_q, num_heads, head_dim]  fp8_e4m3
# kv:            [seq_len, head_dim]           fp8_e4m3
# kv_scales:     [seq_len]                     fp32   (per-position dequant scale)
# weights:       [num_q, num_heads]            fp32   (per-head gate weights)
# cu_end:        [num_q]                        int32  (causal end per query)
# sample_logits: [num_q, sample_len]           fp32   (hot-start gate seed)
idx = flashinfer.dsa_indexer_topk(
    sample_logits, q, kv, kv_scales, weights, cu_end, top_k,
    num_buckets=256,      # gate histogram resolution
    cand_cap=32768,       # candidate buffer size (constant, seq_len-independent)
    refresh_every=64,     # tiles between threshold refreshes
)  # -> [num_q, top_k] int32 KV indices
```

## 📊 Results

**Kernel-level validation (B200, real GLM-5.2 indexer tensors, Q=8192, top_k=2048).** We directly compiled and executed the vendored kernel headers and compared the LiteTopk against the dense logits + TopK baseline. The main motivation is the substantial memory overhead of the dense path, which grows rapidly with context length: **8 GB at 256K, 12 GB at 512K, 24 GB at 768K, and 32 GB at 1M**. The fused kernel removes this overhead while preserving exact TopK results.

| seq_len | dense memory overhead | our memory overhead | dense logits + top-k | fused (vendored) | speedup | recall vs dense |
|--------:|----------------------:|--------------------:|---------------------:|-----------------:|--------:|----------------:|
|    256K |                  8 GB |              1.5 GB |            13.04 ms  |       11.95 ms   |  1.09×  |        100.00 % |
|    512K |                 12 GB |              1.1 GB |            25.99 ms  |       21.95 ms   |  1.18×  |        100.00 % |
|    768K |                 24 GB |              0.7 GB |            39.57 ms  |       31.87 ms   |  1.24×  |        100.00 % |
|      1M |                 32 GB |              0.0 GB |            52.64 ms  |       42.06 ms   |  1.25×  |        100.00 % |

- Fused-vs-dense TopK recall is **100.00%** across tested settings.

- In end-to-end GLM-5.2 prefill, LiteTopK achieves **1.2×** speedup at 1M context and **1.1×** speedup at 768K context without loss.

## 🗂️ Files

Follows FlashInfer's `include/` (framework-agnostic) vs `csrc/` (TVM-FFI binding)
separation — no torch headers in `include/`.

| File | Lines | Role |
|------|------:|------|
| `include/flashinfer/dsa_indexer/dsa_indexer_kernels.cuh` | 715 | Fused CUDA kernels (raw pointers, no torch) |
| `include/flashinfer/dsa_indexer/dsa_indexer.cuh` | 525 | Host-side launch helpers |
| `csrc/dsa_indexer/dsa_indexer.cu` | 172 | TVM-FFI launchers (TensorView) |
| `csrc/dsa_indexer/dsa_indexer_jit_binding.cu` | 33 | `TVM_FFI_DLL_EXPORT_TYPED_FUNC` exports |
| `csrc/dsa_indexer/vendor_deep_gemm/` | — | 10 vendored MIT DeepGEMM SM100 headers + LICENSE + README |
| `flashinfer/jit/dsa_indexer.py` | 35 | JIT module generator (`gen_jit_spec` / `build_and_load`) |
| `flashinfer/dsa_indexer.py` | 149 | Python API |
| `tests/attention/test_dsa_indexer.py` | 75 | Recall test (SM100-gated) |
| `flashinfer/__init__.py` | +1 | `from .dsa_indexer import dsa_indexer_topk` |

## 🧪 Tests

`tests/attention/test_dsa_indexer.py::test_dsa_indexer_topk_recall` asserts
`recall > 0.99` against a dense `deep_gemm` fp32 reference over
`num_q ∈ {64, 512}`, `seq_len ∈ {16384, 65536}`, `top_k ∈ {512, 2048}`. The test is
gated to SM100 (skips on non-Blackwell) and `importorskip`s `deep_gemm`.

Built through the official JIT path (`gen_dsa_indexer_module().build_and_load()`)
and validated on B200; recall table above reproduced from real GLM-5.2 tensors.

## 👤 Author

**Yin Ziqi**, Nanyang Technological University (NTU). 

## Summary by CodeRabbit

* **New Features**
  * Added a CUDA-accelerated DSA indexer for approximate top-k retrieval.
  * Added high-level Python APIs for preparing candidates, scanning inputs, selecting results, and running the complete top-k workflow.
  * Exposed the top-k API at the package level.
  * Added support for SM100 GPUs and configurable probing, refresh, and candidate limits.

* **Tests**
  * Added SM100-gated recall tests validating top-k results against reference computations.

* **Documentation**
  * Added documentation and licensing information for bundled GPU acceleration components.

### CODE REVIEW

**Comment by Heisenberg-Yin** [comment]
@yzh119 @bkryu @aleozlx @yongwww @yyihuang — quick update on this PR: the kernel has been reworked and the diff is now considerably smaller. Would appreciate a look when you have time.

@Aalanli — you worked on the DeepSeek sparse indexer in #2814, so you may find this relevant; the approaches differ (that PR accumulates a histogram in the GEMM epilogue over the materialized logits, this one avoids materializing them at all), but the problem area is shared and your eyes would be valuable.

**Smaller footprint.** The change is down to **10 files / +1968** (from 20 files / +2672). Everything now lives in a single self-contained `flashinfer/experimental/dsa_indexer/` package following the existing `flashinfer.experimental.*` convention, and the **vendored DeepGEMM headers are gone** — the kernels compile against the SM100 headers of the installed `deep_gemm` package instead. No `flashinfer/__init__.py` change: the entry point is gated behind `@flashinfer_experimental_api`, so calling it is the opt-in and there is no automatic backend selection.

**Updated kernel.** The API now returns a `status` tensor alongside the indices, so a row the bounded candidate buffer cannot certify is reported explicitly rather than silently returning a wrong answer. Argument validation (`cand_cap` range, prefix length, 16-byte TMA pointer alignment) is checked up front on both the Python and C++ sides.

**Kernel-level speedup** (B200, real GLM-5.2 layer 0, Q=8192, top-2048), against each framework's native dense indexer top-k path:

| S | vLLM 0.23 native | this PR | speedup | SGLang native | this PR | speedup |
|---:|---:|---:|---:|---:|---:|---:|
| 128K | 6.29 ms | 5.78 ms | 1.09× | 7.74 ms | 5.57 ms | 1.39× |
| 256K | 12.67 ms | 10.58 ms | 1.20× | 15.91 ms | 10.14 ms | 1.57× |
| 512K | 25.54 ms | 20.30 ms | 1.26× | 31.88 ms | 19.64 ms | 1.62× |
| 768K | 38.32 ms | 30.87 ms | 1.24× | 48.48 ms | 27.63 ms | 1.75× |
| 1M | 51.21 ms | 38.82 ms | 1.32× | 64.85 ms | 37.78 ms | 1.72× |

The two "this PR" columns are the same kernel measured inside each framework's own harness, so they differ slightly; the speedup in each pair is against that framework's own baseline.

`benchmarks/bench_dsa_indexer_topk.py` reproduces the latency and memory comparison against the dense path.
